base/compiler/tfuncs.jl

```diff
@@ -1425,8 +1443,38 @@ function tuple_tfunc(argtypes::Vector{Any})
     return anyinfo ? PartialStruct(typ, argtypes) : typ
 end
 
-function arrayref_tfunc(@nospecialize(boundscheck), @nospecialize(a), @nospecialize i...)
-    a = widenconst(a)
+function arrayref_tfunc(@nospecialize(boundcheck), @nospecialize(ary),
+    @nospecialize idxs...)
+    isempty(idxs) && return Bottom
+    array_builtin_common_errorcheck(boundcheck, ary, idxs) || return Bottom
+    return array_elmtype(ary)
+end
+add_tfunc(arrayref, 3, INT_INF, arrayref_tfunc, 20)
+add_tfunc(const_arrayref, 3, INT_INF, arrayref_tfunc, 20)
+
+function arrayset_tfunc(@nospecialize(boundcheck), @nospecialize(ary), @nospecialize(item),
+    @nospecialize idxs...)
+    isempty(idxs) && return Bottom
+    array_builtin_common_errorcheck(boundcheck, ary, idxs) || return Bottom
+    hasintersect(widenconst(item), array_elmtype(ary)) || return Bottom
+    return ary
+end
+add_tfunc(arrayset, 4, INT_INF, arrayset_tfunc, 20)
+
+function array_builtin_common_errorcheck(@nospecialize(boundcheck), @nospecialize(ary),
+    @nospecialize idxs)
+    hasintersect(widenconst(boundcheck), Bool) || return false
+    hasintersect(widenconst(ary), Array) || return false
+    for i = 1:length(idxs)
+        idx = idxs[i]
```

Comment:
```suggestion
        idx = getfield(idxs, i)
```

Reply:
Why don't we want to rely on `getindex` abstraction here?